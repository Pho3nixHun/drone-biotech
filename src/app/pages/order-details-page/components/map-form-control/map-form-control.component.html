@let vM = vm();

<div
    appMapInitializer
    [signal]="mapSignal"
    [mapOptions]="mapOptions"
    style="height: 500px"
    class="mb-4"
></div>

@if (mapSignal(); as map) {
    <app-area-form-control
        [map]="map"
        [formControl]="mapFormGroup.controls.targetAreas"
    />
}

@if (missionStats(); as stats) {
    <app-info-item-list
        class="*:flex-row *:space-y-0 *:justify-between"
        *transloco="let t"
    >
        <app-info-item class="*:flex-col">
            <dt class="text-gray-700">{{ t(vM.totalMissionsLabelKey) }}</dt>
            <dd class="font-semibold">
                {{
                    t(vM.totalMissionsValueKey, { count: stats.totalMissions })
                }}
            </dd>
        </app-info-item>
        <app-info-item class="*:flex-col">
            <dt class="text-gray-700">{{ t(vM.completedMissionsLabelKey) }}</dt>
            <dd class="font-semibold">
                {{
                    t(vM.completedMissionsValueKey, {
                        count: stats.completedMissions,
                    })
                }}
            </dd>
        </app-info-item>
        <app-info-item class="*:flex-col">
            <dt class="text-gray-700">{{ t(vM.remainingMissionsLabelKey) }}</dt>
            <dd class="font-semibold">
                {{
                    t(vM.remainingMissionsValueKey, {
                        count: stats.remainingMissions,
                    })
                }}
            </dd>
        </app-info-item>
    </app-info-item-list>
}
