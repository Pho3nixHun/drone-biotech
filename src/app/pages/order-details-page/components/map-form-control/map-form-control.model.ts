export interface MissionStats {
    totalMissions: number;
    completedMissions: number;
    remainingMissions: number;
}

export interface MapFormControlVM {
    totalMissionsLabelKey: string;
    totalMissionsValueKey: string;
    completedMissionsLabelKey: string;
    completedMissionsValueKey: string;
    remainingMissionsLabelKey: string;
    remainingMissionsValueKey: string;
}
