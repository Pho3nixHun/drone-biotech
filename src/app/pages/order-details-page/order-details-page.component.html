@if (vm(); as vm) {
    <app-page-layout *transloco="let t">
        <app-header>
            <h2>
                {{
                    t(vm.headerXVM.idTitleKey, {
                        id: vm.headerXVM.id,
                    })
                }}
            </h2>

            <app-status [vm]="vm.headerXVM.statusXVM">
                {{ t(vm.headerXVM.statusXVM.statusTextKey) }}
            </app-status>

            <a [routerLink]="vm.headerXVM.addNewMissionsButtonXVM.routerLink">
                <app-button [vm]="vm.headerXVM.addNewMissionsButtonXVM">
                    @if (vm.headerXVM.addNewMissionsButtonXVM.textKey) {
                        {{ t(vm.headerXVM.addNewMissionsButtonXVM.textKey) }}
                    }
                </app-button>
            </a>

            <app-summary-list>
                @for (
                    summary of vm.headerXVM.summaryListXVM.summaryXVMs;
                    track $index
                ) {
                    <app-summary>
                        <p class="text-gray-600">
                            {{ t(summary.textKey) }}
                        </p>

                        <p class="font-medium">
                            @if (summary.value.type === 'withValueKey') {
                                {{ t(summary.value.key, summary.value.params) }}
                            } @else {
                                {{ summary.value.value }}
                            }
                        </p>
                    </app-summary>
                }
            </app-summary-list>
        </app-header>

        @for (sectionCardXVM of vm.sectionCardXVMs; track $index) {
            @switch (sectionCardXVM.type) {
                @case ('orderDetails') {
                    <app-section-card [vm]="sectionCardXVM" *transloco="let t">
                        <div
                            class="flex lg:flex-row flex-col space-y-8 lg:space-y-0"
                        >
                            @for (
                                infoPanelXVM of sectionCardXVM.infoPanelXVMs;
                                track $index
                            ) {
                                <app-info-panel class="lg:w-1/2 w-full">
                                    <h4 class="font-medium">
                                        {{ t(infoPanelXVM.titleKey) }}
                                    </h4>
                                    <app-info-item-list>
                                        @for (
                                            infoItemXVM of infoPanelXVM
                                                .infoListXVM.infoItemXVMs;
                                            track $index
                                        ) {
                                            <app-info-item>
                                                <dt class="text-gray-600 w-32">
                                                    {{
                                                        t(infoItemXVM.labelKey)
                                                    }}
                                                </dt>
                                                <dd>
                                                    <strong class="font-medium">
                                                        @if (
                                                            infoItemXVM.value
                                                                .type ===
                                                            'withValueKey'
                                                        ) {
                                                            {{
                                                                t(
                                                                    infoItemXVM
                                                                        .value
                                                                        .key,
                                                                    infoItemXVM
                                                                        .value
                                                                        .params
                                                                )
                                                            }}
                                                        } @else {
                                                            {{
                                                                infoItemXVM
                                                                    .value.value
                                                            }}
                                                        }
                                                    </strong>
                                                </dd>
                                            </app-info-item>
                                        }
                                    </app-info-item-list>
                                </app-info-panel>
                            }
                        </div>
                    </app-section-card>
                }
                @case ('orderActions') {
                    <app-section-card [vm]="sectionCardXVM" *transloco="let t">
                        <div class="join">
                            <a
                                [href]="
                                    sectionCardXVM.completionTemplateButtonXVM
                                        .link.href
                                "
                                download
                            >
                                <app-button
                                    [vm]="
                                        sectionCardXVM.completionTemplateButtonXVM
                                    "
                                >
                                    @if (
                                        sectionCardXVM
                                            .completionTemplateButtonXVM.textKey
                                    ) {
                                        {{
                                            t(
                                                sectionCardXVM
                                                    .completionTemplateButtonXVM
                                                    .textKey
                                            )
                                        }}
                                    }
                                </app-button>
                            </a>

                            <app-button
                                [vm]="sectionCardXVM.closeOrderButtonXVM"
                                (click)="confirmationDialog.openDialog()"
                            >
                                @if (
                                    sectionCardXVM.closeOrderButtonXVM.textKey
                                ) {
                                    {{
                                        t(
                                            sectionCardXVM.closeOrderButtonXVM
                                                .textKey
                                        )
                                    }}
                                }
                            </app-button>
                        </div>
                    </app-section-card>
                }
                @case ('messages') {
                    <app-section-card [vm]="sectionCardXVM" *transloco="let t">
                        <app-message-item-list>
                            @for (
                                messageItemXVM of sectionCardXVM
                                    .messageItemListXVM.messageItemXVMs;
                                track $index
                            ) {
                                <app-message-item>
                                    @switch (messageItemXVM.avatarXVM.type) {
                                        @case ('withInitials') {
                                            <app-avatar
                                                [class]="
                                                    messageItemXVM.avatarXVM
                                                        .colors
                                                "
                                            >
                                                {{
                                                    messageItemXVM.avatarXVM
                                                        .initials
                                                }}
                                            </app-avatar>
                                        }

                                        @case ('withImage') {
                                            <app-avatar
                                                class="text-white *:bg-black"
                                            >
                                                <img
                                                    [ngSrc]="
                                                        messageItemXVM.avatarXVM
                                                            .imageSrc
                                                    "
                                                    [alt]="
                                                        messageItemXVM.avatarXVM
                                                            .altTextKey
                                                    "
                                                    class="object-contain rounded-full"
                                                    fill
                                                />
                                            </app-avatar>
                                        }
                                    }
                                    <h3 class="font-medium">
                                        {{
                                            t(messageItemXVM.senderValueKey, {
                                                role: t(
                                                    messageItemXVM.roleTextKey
                                                ),
                                                name: messageItemXVM.senderName,
                                            })
                                        }}
                                    </h3>
                                    <p class="text-gray-700">
                                        {{ messageItemXVM.message }}
                                    </p>
                                    <time class="text-gray-600 text-sm">{{
                                        t(messageItemXVM.sendingDateValueKey, {
                                            date: messageItemXVM.sendingDate,
                                        })
                                    }}</time>
                                </app-message-item>
                            }
                        </app-message-item-list>
                        <div class="w-full flex flex-row">
                            <input
                                type="text"
                                [formControl]="messageFormControl"
                                class="px-2 w-full border border-gray-300"
                            />
                            <app-button
                                [vm]="sectionCardXVM.submitButtonXVM"
                                [disabled]="messageFormControl.invalid"
                                (click)="onSendMessageClick()"
                            >
                                @if (sectionCardXVM.submitButtonXVM.icon) {
                                    <mat-icon>{{
                                        t(sectionCardXVM.submitButtonXVM.icon)
                                    }}</mat-icon>
                                }
                            </app-button>
                        </div>
                    </app-section-card>
                }
            }
        }
    </app-page-layout>

    <app-confirmation-dialog
        #confirmationDialog
        [vm]="vm.confirmationDialogVM"
        (response)="onConfirmationDialogResponse($event)"
    />
}
