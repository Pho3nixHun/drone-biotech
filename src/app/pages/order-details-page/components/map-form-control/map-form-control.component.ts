import {
    Component,
    computed,
    effect,
    forwardRef,
    inject,
    input,
    signal,
} from '@angular/core';
import {
    ControlValueAccessor,
    FormBuilder,
    NG_VALUE_ACCESSOR,
    ReactiveFormsModule,
} from '@angular/forms';
import { Coordinates } from '@stores/location/location.model';
import { noop } from 'rxjs';
import { getCenter } from 'geolib';
import { AreaFormControlComponent } from './components/area-form-control/area-form-control.component';
import { TargetArea } from '../../order-details-page.model';
import { MAP_OPTIONS } from 'src/app/shared/providers/google-maps-provider';
import { MapFormControlVM, MissionStats } from './map-form-control.model';
import { toSignal } from '@angular/core/rxjs-interop';
import { TranslocoModule } from '@jsverse/transloco';
import { InfoItemListComponent } from '@components/info-panel/components/info-item-list/info-item-list.component';
import { InfoItemComponent } from '@components/info-panel/components/info-item-list/components/info-item/info-item.component';
import { MapInitializerDirective } from '@directives/map-initializer/map-initializer.directive';

@Component({
    selector: 'app-map-form-control',
    imports: [
        AreaFormControlComponent,
        ReactiveFormsModule,
        TranslocoModule,
        InfoItemListComponent,
        InfoItemComponent,
        MapInitializerDirective,
    ],
    providers: [
        {
            provide: NG_VALUE_ACCESSOR,
            useExisting: forwardRef(() => MapFormControlComponent),
            multi: true,
        },
    ],
    templateUrl: './map-form-control.component.html',
})
export class MapFormControlComponent implements ControlValueAccessor {
    private readonly fb = inject(FormBuilder);
    protected readonly mapOptions = inject(MAP_OPTIONS);

    public vm = input.required<MapFormControlVM>();

    protected readonly mapSignal = signal<google.maps.Map | null>(null);

    protected readonly mapFormGroup = this.fb.group({
        targetAreas: this.fb.control<TargetArea[] | null>(null),
    });

    private readonly targetAreas = toSignal(
        this.mapFormGroup.controls.targetAreas.valueChanges
    );

    protected readonly missionStats = computed<MissionStats | null>(() => {
        const totalMissions = this.targetAreas();
        if (!totalMissions) return null;

        const remainingMissions = totalMissions.filter(
            (area) => area.type === 'active'
        );

        return {
            totalMissions: totalMissions.length,
            remainingMissions: remainingMissions.length,
            completedMissions: totalMissions.length - remainingMissions.length,
        };
    });

    protected readonly mapCenter = computed<Coordinates | null>(() => {
        const areas = this.targetAreas();
        if (!areas) return null;

        const center = getCenter(areas.flatMap((area) => area.coordinates));
        return center
            ? {
                  lat: center.latitude,
                  lng: center.longitude,
              }
            : null;
    });

    protected readonly setMapCenterEffect = effect(() => {
        const map = this.mapSignal();
        const center = this.mapCenter();
        if (!map || !center) return;
        map.setCenter(center);
        this.setMapCenterEffect.destroy();
    });

    public writeValue(targetAreas: TargetArea[]) {
        this.mapFormGroup.setValue({ targetAreas });
    }

    private onChange: (value: TargetArea[]) => void = () => noop;

    private onTouched: () => void = () => noop;

    public registerOnChange(fn: () => void): void {
        this.onChange = fn;
    }

    public registerOnTouched(fn: () => void): void {
        this.onTouched = fn;
    }
}
