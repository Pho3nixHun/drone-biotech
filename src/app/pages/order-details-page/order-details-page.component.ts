import {
    ConfirmationDialogReason,
    ConfirmationDialogVM,
} from '@components/confirmation-dialog/confirmation-dialog.model';
import { FormBuilder, ReactiveFormsModule, Validators } from '@angular/forms';
import { toSignal } from '@angular/core/rxjs-interop';
import { Component, inject, signal } from '@angular/core';
import { NgClass, NgOptimizedImage } from '@angular/common';
import { TranslocoModule } from '@jsverse/transloco';
import { firstValueFrom } from 'rxjs';
import { Store } from '@ngrx/store';
import { selectUser } from '@stores/auth/auth.selector';
import { DialogService } from '@services/dialog/dialog.service';
import { SummaryListComponent } from '@components/summary-list/summary-list.component';
import { StatusComponent } from '@components/status/status.component';
import { SummaryComponent } from '@components/summary-list/components/summary/summary.component';
import { HeaderComponent } from './components/header/header.component';
import { PageLayoutComponent } from '@components/page-layout/page-layout.component';
import { SectionCardComponent } from '@components/section-card/section-card.component';
import { InfoPanelComponent } from '@components/info-panel/info-panel.component';
import { InfoItemListComponent } from '@components/info-panel/components/info-item-list/info-item-list.component';
import { AvatarComponent } from '@components/avatar/avatar.component';
import { InfoItemComponent } from '@components/info-panel/components/info-item-list/components/info-item/info-item.component';
import { ConfirmationDialogComponent } from '@components/confirmation-dialog/confirmation-dialog.component';
import { MessageItemListComponent } from '@components/message-item-list/message-item-list.component';
import { MessageItemComponent } from '@components/message-item/message-item.component';
import { emptyStringValidator } from '@utils/empty-string.validator';
import { OrderDetailsPageService } from './order-details-page.service';
import { GoogleMapComponent } from '@components/google-map/google-map.component';
import { DrawTargetAreasDirective } from '@components/google-map/directives/draw-target-areas/draw-target-areas.directive';
import { CardGroupComponent } from '@components/card-group/card-group.component';
import { CardGroupHeaderComponent } from '@components/card-group/components/card-group-header/card-group-header.component';
import { CardItemListComponent } from '@components/card-group/components/card-item-list/card-item-list.component';
import { CardItemComponent } from '@components/card-group/components/card-item-list/components/card-item/card-item.component';
import { CardItemContentComponent } from '@components/card-group/components/card-item-list/components/card-item/components/card-item-content/card-item-content.component';
import { KeyValueComponent } from '@components/key-value/key-value.component';
import { CardItemActionListComponent } from '@components/card-group/components/card-item-list/components/card-item/components/card-item-action-list/card-item-action-list.component';
import { RouterModule } from '@angular/router';
import { MatIconModule } from '@angular/material/icon';
import { ValueComponent } from '@components/value/value.component';
import { Status } from './order-details-page.model';
import { Avatar } from '@components/avatar/avatar.model';

@Component({
    selector: 'app-order-details-page',
    imports: [
        PageLayoutComponent,
        HeaderComponent,
        TranslocoModule,
        MessageItemListComponent,
        MessageItemComponent,
        StatusComponent,
        SummaryListComponent,
        SummaryComponent,
        SectionCardComponent,
        InfoPanelComponent,
        InfoItemListComponent,
        InfoItemComponent,
        ReactiveFormsModule,
        NgClass,
        AvatarComponent,
        NgOptimizedImage,
        GoogleMapComponent,
        DrawTargetAreasDirective,
        ValueComponent,
        CardGroupComponent,
        CardGroupHeaderComponent,
        CardItemListComponent,
        CardItemComponent,
        CardItemContentComponent,
        KeyValueComponent,
        CardItemActionListComponent,
        RouterModule,
        MatIconModule,
    ],
    templateUrl: './order-details-page.component.html',
})
export class OrderDetailsPageComponent {
    private readonly store = inject(Store);
    private readonly fb = inject(FormBuilder);
    private readonly dialogService = inject(DialogService);
    private readonly orderDetailsPageService = inject(OrderDetailsPageService);

    protected readonly vm = toSignal(this.orderDetailsPageService.getVM());

    protected readonly drawnPolygons = signal<google.maps.Polygon[] | null>(
        null
    );

    protected readonly messageFormControl = this.fb.control('', [
        Validators.required,
        emptyStringValidator(),
    ]);

    protected async onSendMessageClick() {
        const message = this.messageFormControl.value;

        if (this.messageFormControl.invalid || !message) return;

        const sender = await firstValueFrom(this.store.select(selectUser));

        if (!sender) return;

        this.orderDetailsPageService.sendMessage({
            sendingDate: new Date(),
            message,
            sender: {
                role: sender.role,
                name: sender.displayName,
                photoUrl: sender.photoURL,
            },
        });

        this.messageFormControl.reset();
    }

    protected async openConfirmationDialog(vm: ConfirmationDialogVM) {
        const reason: ConfirmationDialogReason = await firstValueFrom(
            this.dialogService.create(vm, ConfirmationDialogComponent).result$
        );

        if (reason.reasonType === 'cancel') return;
        return this.orderDetailsPageService.closeOrder();
    }

    protected readonly Status = Status;
    protected readonly Avatar = Avatar;
}
