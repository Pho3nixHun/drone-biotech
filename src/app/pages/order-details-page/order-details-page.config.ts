import { InjectionToken } from '@angular/core';
import { orderDetailsPageConfig as config } from './order-details-page.mock';
import { OrderDetailsPageConfig } from './order-details-page.model';
import {
    ACTIVE_MISSION_POLYGON_OPTIONS,
    COMPLETED_MISSION_POLYGON_OPTIONS,
} from 'src/app/shared/providers/google-maps-provider';
import { ConfigProvider } from '@interfaces/config-provider';

export const ORDER_DETAILS_PAGE_CONFIG =
    new InjectionToken<OrderDetailsPageConfig>('Order details page config');

export const orderDetailsPageConfig: ConfigProvider[] = [
    {
        provide: ORDER_DETAILS_PAGE_CONFIG,
        useValue: config,
    },
    {
        provide: ACTIVE_MISSION_POLYGON_OPTIONS,
        useValue: {
            clickable: false,
            draggable: false,
            editable: false,
            fillOpacity: 0.5,
            strokeColor: 'green',
            fillColor: 'green',
        },
    },
    {
        provide: COMPLETED_MISSION_POLYGON_OPTIONS,
        useValue: {
            clickable: false,
            draggable: false,
            editable: false,
            fillOpacity: 0.5,
            strokeColor: 'orange',
            fillColor: 'orange',
        },
    },
];
