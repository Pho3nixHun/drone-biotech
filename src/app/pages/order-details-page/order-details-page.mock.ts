import { OrderDetailsPageConfig } from './order-details-page.model';

export const orderDetailsPageConfig: OrderDetailsPageConfig = {
    headerConfig: {
        idTitleKey: 'OrderDetailsPage.header.idTitle',
        clientTextKey: 'OrderDetailsPage.header.clientText',
        createdDateTextKey: 'OrderDetailsPage.header.createdDateText',
        createdDateValueKey: 'OrderDetailsPage.header.createdDateValue',
        totalAreaTextKey: 'OrderDetailsPage.header.totalAreaText',
        totalAreaValueKey: 'OrderDetailsPage.header.totalAreaValue',
        addNewMissionButtonTextKey:
            'OrderDetailsPage.header.addNewMissionButtonText',
    },
    sectionCardConfigs: {
        orderDetailsSectionCardConfig: {
            titleKey: 'OrderDetailsPage.details.title',
            infoPanelConfig: {
                clientInfoPanel: {
                    titleKey: 'OrderDetailsPage.details.client.title',
                    addressLabelKey: 'OrderDetailsPage.details.addressLabel',
                    contactLabelKey: 'OrderDetailsPage.details.contactLabel',
                    emailLabelKey: 'OrderDetailsPage.details.emailLabel',
                    phoneLabelKey: 'OrderDetailsPage.details.phoneLabel',
                },
                summaryInfoPanel: {
                    averageDoseLabelKey:
                        'OrderDetailsPage.details.averageDoseLabel',
                    averageDoseValueKey:
                        'OrderDetailsPage.details.averageDoseValue',
                    orderValueLabelKey:
                        'OrderDetailsPage.details.orderValueLabel',
                    orderValueValueKey:
                        'OrderDetailsPage.details.orderValueValue',
                    titleKey: 'OrderDetailsPage.details.summary.title',
                    totalSupplyLabelKey:
                        'OrderDetailsPage.details.totalSupplyLabel',
                    totalSupplyValueKey:
                        'OrderDetailsPage.details.totalSupplyValue',
                    treatmentLabelKey:
                        'OrderDetailsPage.details.treatmentLabel',
                },
            },
        },
    },
};

export const enMock = {
    headerConfig: {
        idTitle: 'id',
        clientText: 'client',
        createdDateText: 'createdDate',
        createdDateValue: 'createdDateV',
        totalAreaText: 'totalArea',
        totalAreaValue: 'totalAreaV',
        addNewMissionButtonText: 'addNewMissionButton',
    },
    sectionCardConfigs: {
        orderDetailsSectionCardConfig: {
            titleKey: 'title',
            infoPanelConfig: {
                clientInfoPanel: {
                    titleKey: 'title',
                    addressLabelKey: 'address',
                    contactLabelKey: 'contact',
                    emailLabelKey: 'email',
                    phoneLabelKey: 'phone',
                },
                summaryInfoPanel: {
                    averageDoseLabelKey: 'dose',
                    averageDoseValueKey: 'avg',
                    orderValueLabelKey: 'orderL',
                    orderValueValueKey: 'orderV',
                    titleKey: 'title',
                    totalSupplyLabelKey: 'supplyL',
                    totalSupplyValueKey: 'supplyV',
                    treatmentLabelKey: 'treatment',
                },
            },
        },
    },
};

export const orderDetailsPageMockConfig: OrderDetailsPageConfig = {
    headerConfig: {
        idTitleKey: enMock.headerConfig.idTitle,
        clientTextKey: enMock.headerConfig.clientText,
        createdDateTextKey: enMock.headerConfig.createdDateText,
        createdDateValueKey: enMock.headerConfig.createdDateValue,
        totalAreaTextKey: enMock.headerConfig.totalAreaText,
        totalAreaValueKey: enMock.headerConfig.totalAreaValue,
        addNewMissionButtonTextKey: enMock.headerConfig.addNewMissionButtonText,
    },
    sectionCardConfigs: {
        orderDetailsSectionCardConfig: {
            titleKey: 'OrderDetailsPage.details.title',
            infoPanelConfig: {
                clientInfoPanel: {
                    titleKey: 'title',
                    addressLabelKey: 'addressLabel',
                    contactLabelKey: 'contactLabel',
                    emailLabelKey: 'emailLabel',
                    phoneLabelKey: 'phoneLabel',
                },
                summaryInfoPanel: {
                    averageDoseLabelKey: 'averageDoseLabel',
                    averageDoseValueKey: 'averageDoseValue',
                    orderValueLabelKey: 'orderValueLabel',
                    orderValueValueKey: 'orderValueValue',
                    titleKey: 'summary.title',
                    totalSupplyLabelKey: 'totalSupplyLabel',
                    totalSupplyValueKey: 'totalSupplyValue',
                    treatmentLabelKey: 'treatmentLabel',
                },
            },
        },
    },
};
