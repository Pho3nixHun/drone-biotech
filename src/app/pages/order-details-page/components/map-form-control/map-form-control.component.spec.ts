import { ComponentFixture, TestBed } from '@angular/core/testing';

import { MapFormControlComponent } from './map-form-control.component';
import {
    provideActiveMissionPolygonOptions,
    provideCompletedMissionPolygonOptions,
    provideMockMapOptions,
} from 'src/app/shared/providers/google-maps-provider';
import { Component, inject } from '@angular/core';
import { FormBuilder, ReactiveFormsModule } from '@angular/forms';
import { TargetArea } from '../../order-details-page.model';
import { MapFormControlVM } from './map-form-control.model';
import { By } from '@angular/platform-browser';
import { getTranslocoModule } from 'transloco-testing.module';

const enMock = {
    completedMissionsLabel: 'completed',
    completedMissionsValue: 'completedM',
    remainingMissionsLabel: 'remainingM',
    remainingMissionsValue: 'remaining',
    totalMissionsLabel: 'total',
    totalMissionsValue: 'totalM',
};

const vm: MapFormControlVM = {
    completedMissionsLabelKey: enMock.completedMissionsLabel,
    completedMissionsValueKey: enMock.completedMissionsValue,
    remainingMissionsLabelKey: enMock.remainingMissionsLabel,
    remainingMissionsValueKey: enMock.remainingMissionsValue,
    totalMissionsLabelKey: enMock.totalMissionsLabel,
    totalMissionsValueKey: enMock.totalMissionsValue,
};

@Component({
    imports: [MapFormControlComponent, ReactiveFormsModule],
    template: `
        <app-map-form-control [formControl]="targetAreasControl" [vm]="vm" />
    `,
})
class TestHostComponent {
    private readonly fb = inject(FormBuilder);
    public targetAreasControl = this.fb.control<TargetArea[] | null>(null);
    protected vm = vm;
}

describe('MapFormControlComponent', () => {
    let hostComponent: TestHostComponent;
    let fixture: ComponentFixture<TestHostComponent>;
    let innerComponent: MapFormControlComponent;
    let compiled: HTMLElement;

    beforeEach(async () => {
        await TestBed.configureTestingModule({
            imports: [
                TestHostComponent,
                getTranslocoModule({
                    langs: { en: enMock },
                    translocoConfig: {
                        availableLangs: ['en'],
                        defaultLang: 'en',
                    },
                }),
            ],
            providers: [
                provideMockMapOptions(),
                provideActiveMissionPolygonOptions(),
                provideCompletedMissionPolygonOptions(),
            ],
        }).compileComponents();

        fixture = TestBed.createComponent(TestHostComponent);
        hostComponent = fixture.componentInstance;
        innerComponent = fixture.debugElement.query(
            By.directive(MapFormControlComponent)
        ).componentInstance;
        compiled = fixture.debugElement.nativeElement;
    });

    // Unit testing
    it('should set the value of the inner form control', () => {
        // Arrange
        const areas: TargetArea[] = [
            { coordinates: [{ lat: 0, lng: 1 }], type: 'active' },
            { coordinates: [{ lat: 0, lng: 1 }], type: 'active' },
        ];
        hostComponent.targetAreasControl.setValue(areas);

        // Act
        fixture.detectChanges();

        // Assert
        expect(innerComponent['mapFormGroup'].controls.targetAreas.value).toBe(
            areas
        );
    });

    // Unit testing
    it('should calculate the mapCenter', () => {
        // Arrange
        const areas: TargetArea[] = [
            { coordinates: [{ lat: 0, lng: 1 }], type: 'active' },
            { coordinates: [{ lat: 0, lng: 1 }], type: 'active' },
        ];
        hostComponent.targetAreasControl.setValue(areas);

        // Act
        fixture.detectChanges();

        // Assert
        expect(innerComponent['mapCenter']()).toStrictEqual({ lat: 0, lng: 1 });
    });

    // Unit testing
    it('should calculate the stats correctly if there are no targetAreas', () => {
        // There is no need to arrange

        // Act
        fixture.detectChanges();

        // Assert
        expect(innerComponent['missionStats']()).toBe(null);
    });

    // Unit testing
    it('should calculate the stats correctly if there are targetAreas', () => {
        // Arrange
        const areas: TargetArea[] = [
            { coordinates: [{ lat: 0, lng: 1 }], type: 'active' },
            { coordinates: [{ lat: 0, lng: 1 }], type: 'active' },
            { coordinates: [{ lat: 0, lng: 1 }], type: 'completed' },
            { coordinates: [{ lat: 0, lng: 1 }], type: 'completed' },
            { coordinates: [{ lat: 0, lng: 1 }], type: 'active' },
        ];
        hostComponent.targetAreasControl.setValue(areas);

        // Act
        fixture.detectChanges();

        // Assert
        expect(innerComponent['missionStats']()).toStrictEqual({
            completedMissions: 2,
            remainingMissions: 3,
            totalMissions: 5,
        });
    });

    // Snapshot testing
    it('should render the template correctly if there are stats about the targetAreas', () => {
        // Arrange
        const areas: TargetArea[] = [
            { coordinates: [{ lat: 0, lng: 1 }], type: 'active' },
            { coordinates: [{ lat: 0, lng: 1 }], type: 'active' },
        ];
        hostComponent.targetAreasControl.setValue(areas);

        // Act
        fixture.detectChanges();

        // Assert
        expect(compiled).toMatchSnapshot();
    });

    // Snapshot testing
    it('should render the template correctly if there are no stats about the targetAreas', () => {
        // There is no need to arrange

        // Act
        fixture.detectChanges();

        // Assert
        expect(compiled).toMatchSnapshot();
    });
});
