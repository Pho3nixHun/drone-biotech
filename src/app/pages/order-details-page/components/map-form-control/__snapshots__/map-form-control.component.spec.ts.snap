// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapFormControlComponent should render the template correctly if there are no stats about the targetAreas 1`] = `
<div
  id="root5"
  ng-version="19.2.9"
>
  <app-map-form-control
    class="ng-untouched ng-pristine ng-valid"
    ng-reflect-form="[object Object]"
    ng-reflect-vm="[object Object]"
  >
    <div
      appmapinitializer=""
      class="mb-4"
      ng-reflect-map-options="[object Object]"
      ng-reflect-signal="[Signal: null]"
      style="height: 500px;"
    />
    <app-area-form-control
      class="ng-untouched ng-pristine ng-valid"
      ng-reflect-form="[object Object]"
      ng-reflect-map="[object Object]"
    />
    <!--container-->
    <!--container-->
  </app-map-form-control>
</div>
`;

exports[`MapFormControlComponent should render the template correctly if there are stats about the targetAreas 1`] = `
<div
  id="root4"
  ng-version="19.2.9"
>
  <app-map-form-control
    class="ng-untouched ng-pristine ng-valid"
    ng-reflect-form="[object Object]"
    ng-reflect-vm="[object Object]"
  >
    <div
      appmapinitializer=""
      class="mb-4"
      ng-reflect-map-options="[object Object]"
      ng-reflect-signal="[Signal: null]"
      style="height: 500px;"
    />
    <app-area-form-control
      class="ng-untouched ng-pristine ng-valid"
      ng-reflect-form="[object Object]"
      ng-reflect-map="[object Object]"
    />
    <!--container-->
    <app-info-item-list
      class="*:flex-row *:space-y-0 *:justify-between"
    >
      <dl
        class="text-sm flex flex-col space-y-2"
      >
        <app-info-item
          class="*:flex-col"
        >
          <div
            class="flex flex-row"
          >
            <dt
              class="text-gray-700"
            >
              total
            </dt>
            <dd
              class="font-semibold"
            >
               totalM 
            </dd>
          </div>
        </app-info-item>
        <app-info-item
          class="*:flex-col"
        >
          <div
            class="flex flex-row"
          >
            <dt
              class="text-gray-700"
            >
              completed
            </dt>
            <dd
              class="font-semibold"
            >
               completedM 
            </dd>
          </div>
        </app-info-item>
        <app-info-item
          class="*:flex-col"
        >
          <div
            class="flex flex-row"
          >
            <dt
              class="text-gray-700"
            >
              remainingM
            </dt>
            <dd
              class="font-semibold"
            >
               remaining 
            </dd>
          </div>
        </app-info-item>
      </dl>
    </app-info-item-list>
    <!--container-->
    <!--container-->
  </app-map-form-control>
</div>
`;
